dstep/translator/MacroDefinition.d

```diff
@@ -0,0 +1,1403 @@
+    {
+        return ExprType("string");
+    }
+
+    override string toString()
+    {
+        import std.format : format;
+
+        return format("StringConcat(substrings = %s)", substrings);
+    }
+}
+
+class StringsSequence : Expression
+{
+    StringLiteral[] literals;
+}
+
+class IndexExpr : Expression
+{
+    Expression subexpr;
+    Expression index;
+
+    override string translate(ref bool[string] imports)
+    {
+        import std.format : format;
+
+        return format(
+            "%s[%s]",
+            subexpr.translate(imports),
+            index.translate(imports));
+    }
+
+    override ExprType guessExprType()
+    {
+        return ExprType(ExprType.kind.unspecified);
+    }
+
+    override string toString()
+    {
+        import std.format : format;
+
+        return format("IndexExpr(subexpr = %s, index = %s)", subexpr, index);
+    }
+}
+
+class CallExpr : Expression
+{
+    Expression expr;
+    Expression[] args;
+
+    override string translate(ref bool[string] imports)
+    {
+        import std.algorithm.iteration : map;
+        import std.format : format;
+        import std.string : join;
+
+        return format(
+            "%s(%s)",
+            expr.translate(imports),
+            map!(a => a.translate(imports))(args).join(", "));
+    }
+
+    override ExprType guessExprType()
+    {
+        return ExprType(ExprType.kind.unspecified);
+    }
+
+    override string toString()
+    {
+        import std.format : format;
+
+        return format("CallExpr(expr = %s, args = %s)", expr, args);
+    }
+}
+
+class DotExpr : Expression
+{
+    Expression subexpr;
+    string identifier;
+
+    override string translate(ref bool[string] imports)
+    {
+        import std.algorithm.iteration : map;
+        import std.format : format;
+        import std.string : join;
+
+        return format(
+            "%s.%s",
+            subexpr.translate(imports),
+            identifier);
+    }
+
+    override ExprType guessExprType()
+    {
+        return ExprType(ExprType.kind.unspecified);
+    }
+
+    override string toString()
+    {
+        import std.format : format;
+
+        return format(
+            "DotExpr(subexpr = %s, identifier = %s)",
+            subexpr,
+            identifier);
+    }
+}
+
+class ArrowExpr : DotExpr
+{
+    override string toString()
+    {
+        import std.format : format;
+
+        return format(
+            "ArrowExpr(subexpr = %s, identifier = %s)",
+            subexpr,
+            identifier);
+    }
+}
+
+class SubExpr : Expression
+{
+    Expression subexpr;
+
+    this (Expression subexpr)
+    {
+        this.subexpr = subexpr;
+    }
+
+    override string translate(ref bool[string] imports)
+    {
+        import std.format : format;
+
+        if (surplus)
+            return format("%s", subexpr.translate(imports));
+        else
+            return format("(%s)", subexpr.translate(imports));
+    }
+
+    bool surplus() const
```

Comment:
Synonym to `unused`/`excessive`.

[1](http://www.merriam-webster.com/dictionary/surplus)
[2](http://www.thesaurus.com/browse/surplus)